src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -58,60 +62,139 @@ public class QueryService : IQueryService
+
+        stopwatch.Restart();
+        IEnumerable<DicomDataset> instanceMetadata = await GetInstanceMetadataAsync(partitionKey, queryExpression, queryResult, responseBuilder.ReturnTags, cancellationToken);
+        stopwatch.Stop();
+        var resultTime = stopwatch.ElapsedMilliseconds;
+        _logger.LogInformation("QueryService performance filterTimeMilliseconds:{FilterTime}, resultTimeMilliseconds:{ResultTime}", filterTime, resultTime);
+
+        var responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
+        return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
+    }
 
-        if (QueryLimit.ContainsComputedTag(queryExpression.IELevel, responseBuilder.ReturnTags))
+    private static bool ReturnNewTagDefaults(int? apiMajorVersion)
+    {
+        bool useNewDefaults = false;
+        if (apiMajorVersion != null && apiMajorVersion >= 2)
         {
-            var versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+            useNewDefaults = true;
+        }
+        return useNewDefaults;
+    }
+
+    // Does not handle retrieving the extendedQueryTag indexes right now. Logs are in place to evaluate it in the future.
+    private async Task<IEnumerable<DicomDataset>> GetInstanceMetadataAsync(
+        int partitionKey,
+        QueryExpression queryExpression,
+        QueryResult queryResult,
+        IReadOnlyCollection<DicomTag> returnTags,
+        CancellationToken cancellationToken)
+    {
+        bool getStudyResponse = false, getSeriesResponse = false, getFullMetadata = false;
 
-            if (queryExpression.IELevel == Messages.ResourceType.Study)
+        IEnumerable<DicomTag> currentResponseTags = returnTags;
+        // are the expected responses available in StudyResult
+        var remainingTags = currentResponseTags.Except(StudyResult.AvailableTags).ToList();
+        if (currentResponseTags.Count() > remainingTags.Count)
+        {
+            getStudyResponse = true;
+        }
+        // are the remaining expected responses available in SeriesResult
+        if (remainingTags.Any())
+        {
+            currentResponseTags = remainingTags;
+            remainingTags = currentResponseTags.Except(SeriesResult.AvailableTags).ToList();
+            if (currentResponseTags.Count() > remainingTags.Count)
+            {
+                getSeriesResponse = true;
+            }
+        }
+        // if still remaining response tags, just get full metadata
+        if (remainingTags.Any())
+        {
+            _logger.LogInformation("QueryService tags returned from full metadata {FullMetadataTags}", string.Join(',', remainingTags));
+            getFullMetadata = true;
+            // exception of computed tags
+            if (QueryLimit.ContainsComputedTag(queryExpression.IELevel, returnTags))
```

Comment:
This block shouldn't be needed, since the `StudyResult` and `SeriesResult` tag sets checked above include the computed tags.

Reply:
Same as above